### DISCUSSION TITLE
Locks/Atomics cost in Rust migration

### DISCUSSION THREAD

**Discussion by sporksmith** [shadow/shadow#2154] [Ideas]
A frequent topic that comes up as we migrate to Rust is that the most straight-forward way to achieve safety is by introducing new atomic/locking operations, which have some cost.

I sketched out a way to translate Shadow's current safety model to Rust without introducing new atomic operations here: https://github.com/sporksmith/objgraph/ . In microbenchmarks, the benefits of using the proposed `RootedRc` (which avoids atomic operations on clone and drop) vs just using `Arc` appear to be [marginal](https://github.com/sporksmith/objgraph/issues/2#issuecomment-1126967984).

Coming from the other direction, I tried an idea @robgjansen proposed before of doing some benchmarking of a full tor simulation with some extra lock and unlock operations to understand how much it ultimately matters to add some in processing each Event or syscall. For each of the numbers below I just did a single trial of the `tor-shadow-minimal-preload` test.

Adding some locks + unlocks in `event_execute`:

```
    for (int i = 0; i < N; ++i) {
        host_lock(event->dstHost);
        host_unlock(event->dstHost);
    }
```

The cost starts to be noticeable around N=30:

```
0:    14.77s
1:    14.68s
10:   14.83s
25:   14.77s
30:   15.04s  
50:   15.12s
100:  15.62s
1000: 21.58s 
```

If we instead add locks + unlocks in `_syscallhandler_pre_syscall`:

```
    for (int i = 0; i < 15; ++i) {
        host_unlock(sys->host);
        host_lock(sys->host);
    }
```

The cost starts to be noticeable around N=15:

```
0:    14.77s 
1:    14.67s
10:   14.71s
15:   14.99s
20:   15.41s
25:   15.12s
30:   15.43s
50:   15.77s
100:  15.84s
1000: 25.60s
```

If we do both, the cost starts be noticeable around N=15:

```
0:  14.77s
10: 14.80s
15: 14.88s
20: 15.19s
```

Based on our current plans I don't *think* we'd be adding 15+ atomic operations per event and syscall, but it's a bit close for comfort.

It's probably worth repeating this experiment to compare N operations on e.g. `Refcell` vs `AtomicRefcell`, since the `Refcell` cost is a floor of how cheap `RootedRefcell` could possibly be.

If we're uncomfortable with the `Refcell` cost, then we need to architect to avoid even that (e.g. getting rid of shared ownership and refcounting where feasible, which would take some refactoring). 

If the "comfort threshold" is substantially higher for `Refcell` vs `AtomicRefcell` then it may be worth extending the [objgraph](https://github.com/sporksmith/objgraph) idea to add `RootedRefcell` and benchmark that as well.

**Comment by stevenengler** [comment]
I haven't looked at your objgraph code yet, but why are you benchmarking `host_lock()` in this case? Since `Arc` and `AtomicRefCell` are just atomics, wouldn't it be more comparable to benchmark only atomic operations here instead? Maybe the GMutex in `host_lock()` is using atomics in uncontended cases anyways, but I think it would be clearer using just atomics.

**Reply by sporksmith** [reply]
Yeah, it was just convenient for a quick and dirty measurement. My expectation is that lock and unlock should be a single atomic operation, but I could be wrong.

Since a more careful measurement is needed I think I'm going to write a basic wrapper around `Refcell`  and `AtomicRefcell` that we can check here instead, since that's ultimately closer to what we care about. (And/or `Rc` and `Arc`)

**Reply by sporksmith** [reply]
Actually we can test the AtomicRefcell inside `TaskRef::execute` now, which is executed roughly once per event.

```
        for _ in 0..1000 {
            let _ = self.inner.borrow_mut();
        }
```

```
0:     14.77s
25: 14.86s 
50:  14.69s
75: 14.94s
100:  15.17s
1000: 18.33s
```

Here we don't see much change until around 75 operations, suggesting that it is indeed a bit cheaper than the `GMutex`, and seems like a more reasonable amount of breathing room.

**Reply by sporksmith** [reply]
I implemented `RootedRefCell`, which takes [half as long](https://github.com/sporksmith/objgraph/#performance) to lock and unlock than `AtomicRefCell`.

Added this inside `TaskRef::execute`:

```
        use objgraph::{Root, RootedRc, RootedRefCell};
        // In real usage this would hold the Host or be owned by the Host.
        let root = Root::new(());
        // This lock would be taken about the same place as the Host lock (and possibly replace it),
        //  and then be passed around in a context object.
        let lock = root.lock();
        // This is what the inner object of TaskRef would look like.
        let alt_inner = RootedRc::<(), _>::new(root.tag(), RootedRefCell::<(), _>::new(root.tag(), ()));
        for _ in 0..50 {
            let dummy = alt_inner.borrow_mut(&lock);
            // Ensure the borrow  can't be optimized away.
            unsafe { asm!("/* {0} */", in(reg) &dummy) }
        }
        // RootedTaskRef and RootedRc require their associated host locks be held when dropped.
        // (and otherwise panic)
        drop(alt_inner);
```

Comparing in the tor-shadow-minimal benchmark again, this time using hyperfine to run multiple trials. The numbers are slightly higher than above across the board, I think because I'm now reporting the total time for `setup test --extra tor-minimal-shadow-preload` to run instead of just shadow's execution time.

| # Locks  Releases | AtomicRefCell | RootedRefCell |
---------------------------|--------------------|--------------------|
| 0 | 14.966 s ±  0.106 s | - |
| 50 | 15.173 s ±  0.193 s | 15.200 s ±  0.285 s |
| 100 | 15.279 s ±  0.143 s | 15.193 s ±  0.168 s |
| 1000 | 18.309 s ±  0.153 s | 15.846 s ±  0.123 s | 

I still think we can probably carry on using `Arc` and `AtomicRefCell` without worrying about it too much, though if we end up using their atomic operations in the hot path enough to matter, we could get some benefit by in `RootedRc` and `RootedRefCell` (or refactoring).

**Comment by stevenengler** [comment]
Here are some preliminary comments, I only looked briefly at the `RootedRefCell`, and I didn't look into the memory safety very much. I think relying on the internals of `RefCell` and `Rc` is a bit scary, but it's not that bad for our use case. Maybe we could just copy-paste the existing versions of `RefCell` and `Rc` into this code and reduce the number of methods so that we can ensure it's safe.

https://paste.debian.net/hidden/44f6fc02/

**Reply by sporksmith** [reply]
Thanks for the thoughtful review! It's too bad GH doesn't really provide a way to review code that's not in PR form.

I don't have time to through all of it right now, but a couple quick thoughts - 

> relying on the internals of RefCell and Rc is a bit scary

Yeah, agreed. I *think* the RefCell usage is actually safe since we never access it without the lock held.

For `Rc` we do use its `Deref` and `DerefMut` implementations without the lock held, which intuitively *ought* to be safe, but I agree it's better not to rely on that assumption. The easiest fix would be to just quire the lock to dereference as well. Otherwise yeah, we reimplement/fork `Rc`.

> If any of the code above this panics, it will leak memory. Maybe consider
changing `ManuallyDrop` with an `Option` and `Option::take()` instead, which
allows you to do the same thing but without `unsafe` and without the
possibility of memory leaks.

Yeah; panicking in the Drop implementations is pretty unfriendly in general.

I don't think an `Option` works; we really do never want to let the internals be dropped without the lock held.

Still thinking about the best way to handle this. For our purposes leaking is probably ok since I don't think we're ever going to try recovering from panics. If we do publish it as a general purpose crate it'd be nice to have a way to recover safely without leaking. I kind of think we'd need the linked objects (e.g. `RootedRefCell`) to have something like an `AtomicRefCell<RootInternals...>` such that it can push to e.g. a `Vec<Box<dyn Drop>>` that gets dropped when the Root does (or when some method like `Root::gc` is called).

**Reply by stevenengler** [reply]
Also for `RootedRefCell`, instead of holding the mutex guard, what about having a thread-local `borrow_count`? Borrowing increments it, while dropping the `RootedRefCellRef` decrements it. Then have the lock check the borrow count to make sure it's zero before unlocking, otherwise panic.

**Reply by sporksmith** [reply]
Nice; yeah I think that could work.
